core/src/main/scala/org.broadinstitute.dsde.workbench.leonardo/models.scala

```diff
@@ -1,19 +1,146 @@
+
+
+final case class MachineType(value: String) extends AnyVal
+
+sealed trait CloudService extends EnumEntry with Product with Serializable {
+  def asString: String
+
+  override def toString: String = asString //Enumeratum's withName function uses `toString` as key for lookup
+}
+object CloudService extends Enum[CloudService]{
+  case object Dataproc extends CloudService {
+    val asString: String = "DATAPROC"
+  }
+  case object GCE extends CloudService {
+    val asString: String = "GCE"
+  }
+
+  override def values: immutable.IndexedSeq[CloudService] = findValues
+}
+
+final case class ClusterInternalId(asString: String) extends AnyVal
+
+sealed trait RuntimeConfig extends Product with Serializable {
+  def cloudService: CloudService
+  def machineType: MachineType
+  def diskSize: Int
+}
+object RuntimeConfig {
+  final case class GceConfig(
+                              machineType: MachineType,
+                              diskSize: Int
+                            ) extends RuntimeConfig {
+    val cloudService: CloudService = CloudService.GCE
+  }
+
+  final case class DataprocConfig(numberOfWorkers: Int,
+                                  masterMachineType: String,
+                                  masterDiskSize: Int, //min 10
+                                  // worker settings are None when numberOfWorkers is 0
+                                  workerMachineType: Option[String] = None,
+                                  workerDiskSize: Option[Int] = None, //min 10
+                                  numberOfWorkerLocalSSDs: Option[Int] = None, //min 0 max 8
+                                  numberOfPreemptibleWorkers: Option[Int] = None) extends RuntimeConfig {
+    val cloudService: CloudService = CloudService.Dataproc
+    val machineType: MachineType = MachineType(masterMachineType)
+    val diskSize: Int = masterDiskSize
+  }
+}
 
 // Information about service accounts used by the cluster
 final case class ServiceAccountInfo(clusterServiceAccount: Option[WorkbenchEmail],
                                     notebookServiceAccount: Option[WorkbenchEmail])
 
 final case class ClusterError(errorMessage: String, errorCode: Int, timestamp: Instant)
+
+sealed trait UserScriptPath extends Product with Serializable {
+  def asString: String
+}
+object UserScriptPath {
+  final case class Http(url: URL) extends UserScriptPath {
+    val asString: String = url.toString
+  }
+  final case class Gcs(gcsPath: GcsPath) extends UserScriptPath {
+    val asString: String = gcsPath.toUri
+  }
+
+  def stringToUserScriptPath(string: String): Either[Throwable, UserScriptPath] =
+    parseGcsPath(string) match {
+      case Right(value) => Right(Gcs(value))
+      case Left(_)      => Either.catchNonFatal(new URL(string)).map(url => Http(url))
+    }
+}
+
+final case class UserJupyterExtensionConfig(nbExtensions: Map[String, String] = Map.empty,
+                                      serverExtensions: Map[String, String] = Map.empty,
```

Comment:
May need scalafmt in `core` module

Reply:
I think `scalafmt` will apply to `core`. I jsut haven't run it yet :)